clients/src/main/java/org/apache/kafka/clients/ApiVersions.java

```diff
@@ -34,15 +34,27 @@ public class ApiVersions {
 
     private final Map<String, NodeApiVersions> nodeApiVersions = new HashMap<>();
     private byte maxUsableProduceMagic = RecordBatch.CURRENT_MAGIC_VALUE;
+    private short maxSupportedProduceVersion = ApiKeys.PRODUCE.latestVersion();
 
     public synchronized void update(String nodeId, NodeApiVersions nodeApiVersions) {
         this.nodeApiVersions.put(nodeId, nodeApiVersions);
         this.maxUsableProduceMagic = computeMaxUsableProduceMagic();
+        this.maxSupportedProduceVersion = computeMaxSupportedProduceVersion();
+    }
+
+    private short computeMaxSupportedProduceVersion() {
```

Comment:
There is a bit of chicken-and-egg problem. nodeApiVersions is only populated after the client receives the initial ApiVersionResponse during connection creation. However, when the sender generates a produce request, it's possible that the connection to the leader hasn't been established yet. So the sender doesn't know the full nodeApiVersions yet. 

We already have the following generic way of selecting the version of a request in NetworkClient. Perhaps that's enough.

```
            NodeApiVersions versionInfo = apiVersions.get(nodeId);
            short version;
            // Note: if versionInfo is null, we have no server version information. This would be
            // the case when sending the initial ApiVersionRequest which fetches the version
            // information itself.  It is also the case when discoverBrokerVersions is set to false.
            if (versionInfo == null) {
                version = builder.latestAllowedVersion();
                if (discoverBrokerVersions && log.isTraceEnabled())
                    log.trace("No version information found when sending {} with correlation id {} to node {}. " +
                            "Assuming version {}.", clientRequest.apiKey(), clientRequest.correlationId(), nodeId, version);
            } else {
                version = versionInfo.latestUsableVersion(clientRequest.apiKey(), builder.oldestAllowedVersion(),
                        builder.latestAllowedVersion());
            }

```

Reply:
We don't need the supported version on leader per-say here as we just need the supported version cross the nodes we know at the moment not the version on the leader. This why I followed the patter of `maxUsableProduceMagic` here which also use `nodeApiVersions`.  We need the version for backward compatibility between producer and broker version outside NetworkClient. 

Why this chicken-and-egg problem don't apply for `maxUsableProduceMagic`?

Reply:
> Why this chicken-and-egg problem doesn't apply for maxUsableProduceMagic?

That's a good question. We introduced maxUsableProduceMagic for the following reason. 
```
        // use a magic version which is supported by all brokers to reduce the chance that
        // we will need to convert the messages when they are ready to be sent.
```
So, it's used for performance optimization, but not for correctness. It's possible that we miss a node when calculating maxUsableProduceMagic and therefore use a high magic than a broker can handle. This doesn't affect correctness since we will just down convert the message format when actually sending the produce request.

Back to maxSupportedProduceVersion. It's not used for optimization. Do we need it for correctness? For example, suppose that we have the topicId for all topics, is it ok to just use V12 produce request as long as the receiving broker supports it? Do we really need all other brokers to support V12?

Reply:
I have updated the code now, I also ran `client_compatibility_produce_consume_test` and `compatibility_test_new_broker_test` and both passed locally on my machine